consensus/propagation/have_wants.go

```diff
@@ -112,11 +113,155 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
+		return hc, 0
+	}
+	peerConcurrentRequestsCount := peerState.requestCount.Load()
+	_, parts, fullReqs, has := blockProp.getAllState(height, round, false)
+	if !has {
+		blockProp.Logger.Error("couldn't find proposal when filtering requests", "height", height, "round", round)
+		// blockProp.Switch.StopPeerForError(blockProp.getPeer(peer).peer, errors.New("received part for unknown proposal"))
+		// maybe return error
+		return hc, peerConcurrentRequestsCount
+	}
+	perPartLimit := perPartRequestLimit(int(parts.Total()))
+	// TODO also limit per block
+
+	var remainingRequests *bits.BitArray
+
+	// shuffle the indices not to request the parts in the same order
+	shuffledTrueIndices := shuffle(hc.GetTrueIndices())
+	for index, partIndex := range shuffledTrueIndices {
+		reqs := blockProp.countRequests(height, round, partIndex)
+		if len(reqs) >= perPartLimit {
+			hc.SetIndex(partIndex, false)
+			// mark the part as requested.
+			fullReqs.SetIndex(partIndex, true)
+			continue
+		}
+		// don't request the part from this peer if we've already requested it
+		// from them.
+		for _, p := range reqs {
+			// p == peer means we have already requested the part from this peer.
+			if p == peer {
+				hc.SetIndex(partIndex, false)
+				continue
+			}
+		}
+		// stop if we reached the limit of concurrent per peer requests, remove
+		// the remaining indexes and save them in the remaining parts.
+		if peerConcurrentRequestsCount >= concurrentPerPeerRequestLimit {
+			remainingRequests = bits.NewBitArray(hc.Size())
+			for _, partId := range shuffledTrueIndices[index:] {
+				remainingRequests.SetIndex(partId, true)
+				hc.SetIndex(partId, false)
+			}
+			break
+		}
+		peerConcurrentRequestsCount++
+	}
+
+	return remainingRequests, peerConcurrentRequestsCount
+}
+
+func (blockProp *Reactor) enqueueRemainingRequests(peer *PeerState, height int64, round int32, remaining *bits.BitArray) {
+	select {
+	case <-blockProp.ctx.Done():
+		return
+	case peer.remainingRequests <- &remainingRequests{
+		height:   height,
+		round:    round,
+		requests: remaining,
+	}:
+	}
+}
+
+func (blockProp *Reactor) processRemainingRequests(d *PeerState) {
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case remainingReq, has := <-d.remainingRequests:
+			if !blockProp.started.Load() {
+				continue
+			}
+			if !has {
+				return
+			}
+			_, parts, _, has := blockProp.getAllState(remainingReq.height, remainingReq.round, true)
+			if !has {
+				blockProp.Logger.Error("couldn't find proposal when processing remaining requests", "height", remainingReq.height, "round", remainingReq.round)
+				return
+			}
+			// check if we received these parts from anywhere
+			toBeRequested := remainingReq.requests
+			var remainingBA *bits.BitArray
+			concurrentPeerRequestsCount := int64(0)
+			for remainingBA != nil && !remainingBA.IsEmpty() {
+				select {
+				case <-blockProp.ctx.Done():
+					return
+				case _, has := <-d.receivedPart:
```

Comment:
nested for and select statements _almost_ always means that refactoring is needed

Reply:
didn't know that, any idea how to loop until the remaining requests are nil but still stop if the context is done without using a for + select?